### ISSUE TITLE
Error messages on LTI grade sync could be improved

### ISSUE THREAD

**Issue by mwest1066** [PrairieLearn/PrairieLearn#10900]
Some potential improvements to the format of the sync output:
1. The "Could not find LTI" error message could say something more like:
    > Could not find Matthew West as a student in the linked Illinois Canvas course (UID mwest@illinois.edu, email mwest@illinois.edu)
2. The "Unprocessable entity" error messages could be explicitly rendered as:
    > Could not sync grade for Matthew West. This user is not a student in the linked Illinois Canvas course (UID mwest@illinois.edu, email mwest@illinois.edu)
3. Change the default listing for students to include the UID. As @lslavice has pointed out before, we should always include UIDs for students because she's had cases where students have identical names. We should also remove the `ai=...` text unless it's serving some purpose here?

In the example error messages above I'm including the email because I think that's what we are actually trying to link on? The UIDs are included to uniquely identify the student in PL.

Example current output:

![Image](https://github.com/user-attachments/assets/f67a959f-2588-49bb-a47f-af493aee3daa)

**Comment by mfsilva22** [comment]
great improvements! It will for sure reduce unnecessary panic from instructors. The word "errors" makes me think I broke something, which is not the case here.

**Comment by lslavice** [comment]
I agree, these would be a nice format iteration on what we see in the grade sync! And thanks for remembering about me and my duplicately named students -- we have another pair this semester!! hahaha

**Comment by mwest1066** [comment]
We should also keep track of all users whose grades were not synced and print a summary list of them at the end, so that instructors don't need to scroll back looking for problems.

**Comment by trombonekenny** [comment]
I'm working on this and have an updated version:

![Image](https://github.com/user-attachments/assets/530ef047-1c01-4942-968b-69b78a74ad70)

I realized these were essentially the same error case (but happen in different places in the process with different messaging). Should we make them report the same way?

I believe we have enough information to know who is not a student before we start sending grades (from PL or Canvas or both). We could put more logic around that to not send or not display errors for the course staff, but I'm not sure if that's the right way to go. I don't want to be too restrictive or set is up for challenges with other LMSes. Thoughts?

**Comment by mwest1066** [comment]
> I'm working on this and have an updated version:

Nice!

> I realized these were essentially the same error case (but happen in different places in the process with different messaging). Should we make them report the same way?

I think so. From the user point of view these are basically the same thing, so making them consistent would be friendly.

> I believe we have enough information to know who is not a student before we start sending grades (from PL or Canvas or both). We could put more logic around that to not send or not display errors for the course staff, but I'm not sure if that's the right way to go. I don't want to be too restrictive or set is up for challenges with other LMSes. Thoughts?

I'm very much in favor of excluding course staff and not generating errors for them. At the moment the errors from course staff can generate a lot of noise that obscures sync problems for actual students.

I also wouldn't worry too much at this point about other LMSes -- they'll undoubtedly have their own quirks that we'll need to work around anyway.

**Comment by trombonekenny** [comment]
I wanted to recap our conversation from today because I'm not sure if my notes are right.

1) Non-student in PL and errors during pushing, try to push but suppress the error. We don't want to be distracted by staff errors.
2) Student in PL but non-student in Canvas (meaning no learner role), don't even try to push. It would error anyway and we don't care about those.
3) Student in PL but not a student in Canvas, show the error.

**Comment by mwest1066** [comment]
> 1. Non-student in PL and errors during pushing, try to push but suppress the error. We don't want to be distracted by staff errors.

Can we detect this ahead of time and don't try to push? If we have a list of Canvas students we can skip the push if the user is non-student in PL and not in Canvas students?

> 2. Student in PL but non-student in Canvas (meaning no learner role), don't even try to push. It would error anyway and we don't care about those.
> 3. Student in PL but not a student in Canvas, show the error.

These two look good to me.

One other category is non-student in PL and student in Canvas. I think we should push this.

To summarize it, I think we can get three lists of users (tell me if we don't actually have these lists):
* non-students in PL
* students in PL
* students in Canvas

The groups are:
1. non-student in PL and student in Canvas ==> push and report error if it fails
2. non-student in PL and NOT student in Canvas ==> don't try to push
3. student in PL and student in Canvas ==> push and report error if it fails
4. student in PL and NOT student in Canvas ==> report error

I guess my one question is whether we can determine in advance the list of Canvas students, or do we just need to push and find out who errors?

**Comment by trombonekenny** [comment]
Thanks for the summary, @mwest1066. Is
```
      const is_staff = await callRow(
        'users_is_instructor_in_course_instance',
        [user.user_id, assessment.course_instance_id],
        z.boolean(),
      );
```
sufficient for determining non-student in PL, or do you have a different recommendation?

**Comment by mwest1066** [comment]
> Thanks for the summary, [@mwest1066](https://github.com/mwest1066). Is
> 
> sufficient for determining non-student in PL, or do you have a different recommendation?

That's the right way to do it!
